python/pyarrow/_dataset.pyx

```diff
@@ -1421,34 +1471,72 @@ cdef class IsValidExpression(UnaryExpression):
         expr = make_shared[CIsValidExpression](operand.unwrap())
         self.init(<shared_ptr[CExpression]> expr)
 
+    def __reduce__(self):
+        return IsValidExpression, (self.operand,)
+
 
 cdef class CastExpression(UnaryExpression):
 
+    cdef CCastExpression *cast
+
     def __init__(self, Expression operand not None, DataType to not None,
                  bint safe=True):
-        # TODO(kszucs): safe is consistently used across pyarrow, but on long
-        #               term we should expose the CastOptions object
         cdef:
             CastOptions options
             shared_ptr[CExpression] expr
         options = CastOptions.safe() if safe else CastOptions.unsafe()
-        expr.reset(new CCastExpression(
-            operand.unwrap(),
-            pyarrow_unwrap_data_type(to),
-            options.unwrap()
-        ))
+        expr.reset(
+            new CCastExpression(
+                operand.unwrap(),
+                pyarrow_unwrap_data_type(to),
+                options.unwrap()
+            )
+        )
         self.init(expr)
 
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        UnaryExpression.init(self, sp)
+        self.cast = <CCastExpression*> sp.get()
+
+    @property
+    def to(self):
+        # safe to assume that CastExpression::to_ variant holds a DataType
+        # instance because the construction from python only allows that
+        return pyarrow_wrap_data_type(self.cast.to_type())
+
+    @property
+    def safe(self):
+        cdef CCastOptions options = self.cast.options()
+        # infer safeness from any of the allow_* properties of the cast option
+        return not options.allow_int_overflow
+
+    def __reduce__(self):
+        return CastExpression, (self.operand, self.to, self.safe)
+
 
 cdef class InExpression(UnaryExpression):
 
+    cdef:
+        CInExpression* inexpr
+
     def __init__(self, Expression operand not None, Array haystack not None):
         cdef shared_ptr[CExpression] expr
         expr.reset(
             new CInExpression(operand.unwrap(), pyarrow_unwrap_array(haystack))
         )
         self.init(expr)
 
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        UnaryExpression.init(self, sp)
+        self.inexpr = <CInExpression*> sp.get()
+
+    @property
+    def values(self):
```

Comment:
please rename this to set or set_ or something. The identifier `values` is already too overloaded in our codebase

Reply:
The constructor uses `haystack` (but I would also make this a private property for now)

Reply:
Done.